### PR TITLE
[TS migration] Migrate 'SettlementButton.js' component to TypeScript

### PR DESCRIPTION
**Pull request by ZhenjaHorbach**
### Fixed Issues

$ https://github.com/Expensify/App/issues/25100
PROPOSAL: https://github.com/Expensify/App/issues/25100#issuecomment-1860945360

### Tests

1. Go to request money
2. Select the Distance tab
3. Enter any waypoints
4. Select any workspace
5. Create request
6. Make sure SettlementButton works correctly on Report Preview
7.  Press on Report Preview 
8. Make sure SettlementButton works correctly on Money Report Header

- [x] Verify that no errors appear in the JS console

### Offline tests

1. Go to request money
2. Select the Distance tab
3. Enter any waypoints
4. Select any workspace
5. Create request
6. Make sure SettlementButton works correctly on Report Preview
7.  Press on Report Preview 
8. Make sure SettlementButton works correctly on Money Report Header

### QA Steps

1. Go to request money
2. Select the Distance tab
3. Enter any waypoints
4. Select any workspace
5. Create request
6. Make sure SettlementButton works correctly on Report Preview
7.  Press on Report Preview 
8. Make sure SettlementButton works correctly on Money Report Header

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/68128028/0d9cd697-215a-4a45-8bc1-4d1549d4ce3d

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/68128028/0f03e542-3e55-49ce-b47a-15be03a12cab

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/68128028/d90983d8-b211-41a1-b084-2f1ec0db5c22

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/68128028/b7eb5918-a473-4b57-ba17-3d4c760737f1

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/68128028/8cbaa0fc-707b-4f8a-9a68-5d01196221a1

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/68128028/e1b57e20-0347-4c36-8116-a3042547ccff

</details>

### CODE REVIEW

File: src/components/SettlementButton.tsx [hunk 1]
```diff
@@ -1,143 +1,106 @@
...
+import {EmptyObject} from '@src/types/utils/EmptyObject';
 import ButtonWithDropdownMenu from './ButtonWithDropdownMenu';
 import * as Expensicons from './Icon/Expensicons';
 import KYCWall from './KYCWall';
-import withNavigation from './withNavigation';
 
-const propTypes = {
-    /** Callback to execute when this button is pressed. Receives a single payment type argument. */
-    onPress: PropTypes.func.isRequired,
+type AnchorAlignment = {
```
Comment (kubabutkiewicz) [inline src/components/SettlementButton.tsx:21]: this type already exists in [Popover](https://github.com/Expensify/App/blob/main/src/components/Popover/types.ts) component , please reuse it

File: src/components/SettlementButton.tsx [hunk 2]
```diff
@@ -1,143 +1,106 @@
...
 
-    /** Call the onPress function on main button when Enter key is pressed */
-    pressOnEnter: PropTypes.bool,
+type TriggerKYCFlow = (event: Event, iouPaymentType: string) => void;
 
-    /** Settlement currency type */
-    currency: PropTypes.string,
+type SettlementButtonOnyxProps = {
+    /** The last payment method used per policy */
+    nvpLastPaymentMethod?: Record<string, string>;
```
Comment (kubabutkiewicz) [inline src/components/SettlementButton.tsx:30]: this comes from Onyx so it should be typed as `OnyxEntry<Record<string, string>>`

File: src/components/SettlementButton.tsx [hunk 3]
```diff
@@ -1,143 +1,106 @@
...
+};
 
+type SettlementButtonProps = SettlementButtonOnyxProps & {
+    /** Callback to execute when this button is pressed. Receives a single payment type argument. */
+    onPress: (paymentType: string) => void;
```
Comment (kubabutkiewicz) [inline src/components/SettlementButton.tsx:35]: I belive paymentType should be typed as `DeepValueOf<typeof CONST.IOU.PAYMENT_TYPE>`

File: src/components/SettlementButton.tsx [hunk 4]
```diff
@@ -1,143 +1,106 @@
...
+    /** The route to redirect if user does not have a payment method setup */
+    enablePaymentsRoute: string;
```
Comment (kubabutkiewicz) [inline src/components/SettlementButton.tsx:37]: I think this shouldn't be just any string but we should limit this to only few values. I found those used for this prop
`ROUTES.ENABLE_PAYMENTS`, `ROUTES.IOU_SEND_ENABLE_PAYMENTS`, `ROUTES.SETTINGS_ENABLE_PAYMENTS`

File: src/components/SettlementButton.tsx [hunk 5]
```diff
@@ -1,143 +1,106 @@
...
-
+    style?: StyleProp<ViewStyle>;
     /** Total money amount in form <currency><amount> */
-    formattedAmount: PropTypes.string,
-
+    formattedAmount?: string;
     /** The size of button size */
-    buttonSize: PropTypes.oneOf(_.values(CONST.DROPDOWN_BUTTON_SIZE)),
-
+    buttonSize?: ButtonSizeValue; // Replace with the actual sizes
```
Comment (kubabutkiewicz) [inline src/components/SettlementButton.tsx:57]: is this a todo?

File: src/components/SettlementButton.tsx [hunk 6]
```diff
@@ -1,143 +1,106 @@
...
-    shouldHidePaymentOptions,
-    shouldShowApproveButton,
-    style,
-    shouldShowPersonalBankAccountOption,
-}) {
+    pressOnEnter = false,
+    policyID = '',
+    shouldHidePaymentOptions = false,
+    shouldShowApproveButton = false,
+    style = [],
```
Comment (kubabutkiewicz) [inline src/components/SettlementButton.tsx:101]: default value for style props is not necessary

File: src/components/SettlementButton.tsx [hunk 7]
```diff
@@ -227,15 +190,15 @@ function SettlementButton({
...
                 <ButtonWithDropdownMenu
                     buttonRef={buttonRef}
                     isDisabled={isDisabled}
                     isLoading={isLoading}
-                    onPress={(event, iouPaymentType) => selectPaymentType(event, iouPaymentType, triggerKYCFlow)}
+                    onPress={(event: Event, iouPaymentType: string) => selectPaymentType(event, iouPaymentType, triggerKYCFlow)}
                     pressOnEnter={pressOnEnter}
                     options={paymentButtonOptions}
-                    style={style}
+                    style={style as Record<string, string | number>}
```
Comment (kubabutkiewicz) [inline src/components/SettlementButton.tsx:201]: I think this assertion is not needed

File: src/components/SettlementButton.tsx [hunk 8]
```diff
@@ -227,15 +190,15 @@ function SettlementButton({
             anchorAlignment={kycWallAnchorAlignment}
             shouldShowPersonalBankAccountOption={shouldShowPersonalBankAccountOption}
         >
-            {(triggerKYCFlow, buttonRef) => (
+            {(triggerKYCFlow: TriggerKYCFlow, buttonRef: MutableRefObject<null>) => (
```
Comment (kubabutkiewicz) [inline src/components/SettlementButton.tsx:193]: why this ref is typed as `null`?

File: src/components/SettlementButton.tsx (hunk 5 above)
Comment (blazejkustra) [reply src/components/SettlementButton.tsx:58]: I think he meant this comment: "// Replace with the actual sizes"

File: src/components/SettlementButton.tsx (hunk 8 above)
Comment (blazejkustra) [reply src/components/SettlementButton.tsx:194]: If you track it down it should be `HTMLDivElement | View` instead, it's not typed yet as BaseKYCWall is in JS

File: src/components/SettlementButton.tsx [hunk 9]
```diff
@@ -1,143 +1,107 @@
-import PropTypes from 'prop-types';
+import _ from 'lodash';
```
Comment (blazejkustra) [inline src/components/SettlementButton.tsx:1]: Remove lodash from this file, instead use vanilla JS methods if possible

File: src/components/SettlementButton.tsx [hunk 10]
```diff
@@ -1,143 +1,107 @@
...
+import type {EmptyObject} from '@src/types/utils/EmptyObject';
 import ButtonWithDropdownMenu from './ButtonWithDropdownMenu';
 import * as Expensicons from './Icon/Expensicons';
 import KYCWall from './KYCWall';
-import withNavigation from './withNavigation';
 
-const propTypes = {
-    /** Callback to execute when this button is pressed. Receives a single payment type argument. */
-    onPress: PropTypes.func.isRequired,
+type TriggerKYCFlow = (event: Event, iouPaymentType: string) => void;
```
Comment (blazejkustra) [inline src/components/SettlementButton.tsx:25]: Can we make event a little bit more specific than `Event`?

File: src/components/SettlementButton.tsx (hunk 7 above)
Comment (blazejkustra) [reply src/components/SettlementButton.tsx:204]: Use `@ts-expect-error` instead, please refer to TS guidelines

File: src/components/SettlementButton.tsx [hunk 11]
```diff
@@ -1,143 +1,107 @@
...
 
-    /** Call the onPress function on main button when Enter key is pressed */
-    pressOnEnter: PropTypes.bool,
+type PaymentType = DeepValueOf<typeof CONST.IOU.PAYMENT_TYPE | typeof CONST.IOU.REPORT_ACTION_TYPE>;
 
-    /** Settlement currency type */
-    currency: PropTypes.string,
+type SettlementButtonOnyxProps = {
+    /** The last payment method used per policy */
+    nvpLastPaymentMethod?: OnyxEntry<Record<string, string>>;
```
Comment (blazejkustra) [inline src/components/SettlementButton.tsx:31]: Let's create a type for `NVP_LAST_PAYMENT_METHOD` inside src/types/onyx:
```ts
    [ONYXKEYS.NVP_LAST_PAYMENT_METHOD]: Record<string, string>;
```

File: src/components/SettlementButton.tsx [hunk 12]
```diff
@@ -1,143 +1,107 @@
...
+};
 
+type SettlementButtonProps = SettlementButtonOnyxProps & {
+    /** Callback to execute when this button is pressed. Receives a single payment type argument. */
+    onPress: (paymentType: PaymentType) => void;
+    /** The route to redirect if user does not have a payment method setup */
```
Comment (blazejkustra) [inline src/components/SettlementButton.tsx:37]: Add spaces between props, same as in other components

File: src/components/SettlementButton.tsx [hunk 13]
```diff
@@ -244,15 +210,11 @@ function SettlementButton({
...
-export default compose(
-    withNavigation,
-    withOnyx({
-        nvp_lastPaymentMethod: {
-            key: ONYXKEYS.NVP_LAST_PAYMENT_METHOD,
-        },
-    }),
-)(SettlementButton);
+export default withOnyx<SettlementButtonProps, SettlementButtonOnyxProps>({
+    nvpLastPaymentMethod: {
```
Comment (blazejkustra) [inline src/components/SettlementButton.tsx:216]: Good one, thanks for changing it to `nvpLastPaymentMethod`

**Review by blazejkustra — CHANGES_REQUESTED** [review]

**Comment by blazejkustra** [comment]
Also a kind request, change PR title so it reflects issue title: `[TS migration] Migrate 'SettlementButton.js' component to TypeScript`

File: src/components/SettlementButton.tsx [hunk 14]
```diff
@@ -1,143 +1,127 @@
...
+import type {EmptyObject} from '@src/types/utils/EmptyObject';
 import ButtonWithDropdownMenu from './ButtonWithDropdownMenu';
 import * as Expensicons from './Icon/Expensicons';
 import KYCWall from './KYCWall';
-import withNavigation from './withNavigation';
 
-const propTypes = {
+type Event = GestureResponderEvent | KeyboardEvent;
+
+type TriggerKYCFlow = (event: Event, iouPaymentType: string) => void;
```
Comment (blazejkustra) [inline src/components/SettlementButton.tsx:27]: Can we make event a little bit more specific than Event?

File: src/components/SettlementButton.tsx [hunk 15]
```diff
@@ -1,143 +1,127 @@
...
+
+type PaymentType = DeepValueOf<typeof CONST.IOU.PAYMENT_TYPE | typeof CONST.IOU.REPORT_ACTION_TYPE>;
+
+type SettlementButtonOnyxProps = {
+    /** The last payment method used per policy */
+    nvpLastPaymentMethod?: OnyxEntry<OnyxValues[typeof ONYXKEYS.NVP_LAST_PAYMENT_METHOD]>;
+};
```
Comment (blazejkustra) [inline src/components/SettlementButton.tsx:34]: In my previous comment I meant to create a type and use it here like this:

```ts
// src/types/onyx/LastPaymentMethod.ts
type LastPaymentMethod = Record<string, string>;
```

```suggestion
type SettlementButtonOnyxProps = {
    /** The last payment method used per policy */
    nvpLastPaymentMethod?: OnyxEntry<LastPaymentMethod>;
};
```

**Review by blazejkustra — CHANGES_REQUESTED** [review]

File: src/types/onyx/LastPaymentMethod.ts [hunk 16]
```diff
@@ -0,0 +1,3 @@
+type LastPaymentMethod = Record<string, string>;
+
+export default LastPaymentMethod;
```
Comment (blazejkustra) [inline src/types/onyx/LastPaymentMethod.ts:3]: LEt's use this type in ONYXKEYS.ts as well.

File: src/components/SettlementButton.tsx [hunk 17]
```diff
@@ -1,143 +1,125 @@
...
+    nvpLastPaymentMethod?: OnyxEntry<LastPaymentMethod>;
+};
+
+type SettlementButtonProps = SettlementButtonOnyxProps & {
     /** Callback to execute when this button is pressed. Receives a single payment type argument. */
-    onPress: PropTypes.func.isRequired,
+    onPress: (paymentType: PaymentType) => void;
+
+    /** The route to redirect if user does not have a payment method setup */
+    enablePaymentsRoute: typeof ROUTES.ENABLE_PAYMENTS | typeof ROUTES.IOU_SEND_ENABLE_PAYMENTS | typeof ROUTES.SETTINGS_ENABLE_PAYMENTS;
```
Comment (blazejkustra) [inline src/components/SettlementButton.tsx:39]: Wdyt to extract `typeof ROUTES.ENABLE_PAYMENTS | typeof ROUTES.IOU_SEND_ENABLE_PAYMENTS | typeof ROUTES.SETTINGS_ENABLE_PAYMENTS;` to a type? Also did you make sure this type is correct?

File: src/ONYXKEYS.ts [hunk 18]
```diff
@@ -530,5 +530,7 @@ type OnyxValues = {
 
 type OnyxKeyValue<TOnyxKey extends (OnyxKey | OnyxCollectionKey) & keyof OnyxValues> = OnyxEntry<OnyxValues[TOnyxKey]>;
 
+type LastPaymentMethod = Record<string, string>;
```
Comment (blazejkustra) [inline src/ONYXKEYS.ts:533]: Do not define it here, just use it in ONYXKEYS in the OnyxValues type:
```ts
import type * as OnyxTypes from './types/onyx';
type OnyxValues = {
    ...
    [ONYXKEYS.NVP_LAST_PAYMENT_METHOD]: OnyxTypes.LastPaymentMethod;
    ...
}
```

File: src/components/SettlementButton.tsx [hunk 19]
```diff
@@ -1,143 +1,128 @@
...
-import withNavigation from './withNavigation';
 
-const propTypes = {
+type KYCFlowEvent = GestureResponderEvent | KeyboardEvent;
+
+type TriggerKYCFlow = (event: KYCFlowEvent, iouPaymentType: string) => void;
+
+type PaymentType = DeepValueOf<typeof CONST.IOU.PAYMENT_TYPE | typeof CONST.IOU.REPORT_ACTION_TYPE>;
+
+type EnablePaymentsRoute = typeof ROUTES.ENABLE_PAYMENTS | typeof ROUTES.IOU_SEND_ENABLE_PAYMENTS | typeof ROUTES.SETTINGS_ENABLE_PAYMENTS;
```
Comment (blazejkustra) [inline src/components/SettlementButton.tsx:30]: Thank you for this 👍

**Review by blazejkustra — APPROVED** [review]
LGTM, please merge the newest main

**Review by hoangzinh — COMMENTED** [review]
Look good. Small feedback from me

**Review by hoangzinh — APPROVED** [review]
LGTM

**Comment by AndrewGable** [comment]
Looking good, but conflicts. Will give a final review when they are resolved.

**Comment by AndrewGable** [comment]
@hoangzinh - Can you re-review? I've seen a few merges with main that introduce issues and would love to make sure this isn't one of them 😄

**Comment by blazejkustra** [comment]
Please sync this branch with the newest main. Context [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1706196639925099), thanks!

File: src/libs/actions/Policy.ts [hunk 20]
```diff
@@ -1611,12 +1612,14 @@ function buildOptimisticPolicyRecentlyUsedTags(policyID: string, tag: string): R
...
  */
-function createWorkspaceFromIOUPayment(iouReport: Report): string | undefined {
+function createWorkspaceFromIOUPayment(iouReportParam: Report | EmptyObject): string | undefined {
     // This flow only works for IOU reports
-    if (!ReportUtils.isIOUReport(iouReport)) {
+    if (!ReportUtils.isIOUReport(iouReportParam)) {
         return;
     }
 
+    const iouReport = iouReportParam as Report;
```
[the remark this answers is not in the document]
Comment (blazejkustra) [reply src/libs/actions/Policy.ts]: > @ZhenjaHorbach @blazejkustra please you have time, could you give us advice here? I don't really like this way

Me neither haha

> Or is there any way for us to TS know that after call util ReportUtils.isIOUReport, it should be a Report type.

Yes, this should be possible! It's called [user defined type guard](https://levelup.gitconnected.com/user-defined-type-guards-in-typescript-fad639e4944f). We may need to create two different functions because of this (one for passing a report and one for passing reportID)

```ts
function isIOUReport(report: OnyxEntry<Report> | EmptyObject): report is Report {
    return report?.type === CONST.REPORT.TYPE.IOU;
}
```

File: src/components/SettlementButton.tsx [hunk 21]
```diff
@@ -227,12 +213,12 @@ function SettlementButton({
             anchorAlignment={kycWallAnchorAlignment}
             shouldShowPersonalBankAccountOption={shouldShowPersonalBankAccountOption}
         >
-            {(triggerKYCFlow, buttonRef) => (
+            {(triggerKYCFlow: TriggerKYCFlow, buttonRef: ForwardedRef<HTMLElement>) => (
```
Comment (blazejkustra) [inline src/components/SettlementButton.tsx:216]: Why not infer the types here?
```suggestion
            {(triggerKYCFlow, buttonRef) => (
```

File: src/components/SettlementButton.tsx [hunk 22]
```diff
@@ -227,12 +213,12 @@ function SettlementButton({
...
                 <ButtonWithDropdownMenu
-                    buttonRef={buttonRef}
+                    buttonRef={buttonRef as RefObject<View>}
```
Comment (blazejkustra) [inline src/components/SettlementButton.tsx:218]: Please console.log the type of buttonRef here, I suspect the type is wrong in either KYCWall or ButtonWithDropdownMenu

File: src/components/SettlementButton.tsx [hunk 23]
```diff
@@ -227,12 +213,12 @@ function SettlementButton({
...
                     isDisabled={isDisabled}
                     isLoading={isLoading}
-                    onPress={(event, iouPaymentType) => selectPaymentType(event, iouPaymentType, triggerKYCFlow)}
+                    onPress={(event: KYCFlowEvent, iouPaymentType: PaymentType) => selectPaymentType(event, iouPaymentType, triggerKYCFlow)}
```
Comment (blazejkustra) [inline src/components/SettlementButton.tsx:221]: Does it work?
```suggestion
                    onPress={(event, iouPaymentType) => selectPaymentType(event, iouPaymentType, triggerKYCFlow)}
```

File: src/components/KYCWall/types.ts [hunk 24]
```diff
@@ -49,7 +47,7 @@ type KYCWallProps = {
     chatReportID?: string;
 
     /** The IOU/Expense report we are paying */
-    iouReport?: OnyxEntry<Report>;
+    iouReport?: OnyxEntry<Report> | EmptyObject;
```
Comment (blazejkustra) [inline src/components/KYCWall/types.ts:50]: Why is this needed?

**Review by blazejkustra — CHANGES_REQUESTED** [review]

File: src/components/SettlementButton.tsx (hunk 22 above)
Comment (blazejkustra) [reply src/components/SettlementButton.tsx:218]: Shouldn't it be `HTMLElement | View` in `KYCWall` then?

File: src/components/KYCWall/types.ts (hunk 24 above)
Comment (blazejkustra) [reply src/components/KYCWall/types.ts:50]: Looks okay 👍

File: src/components/SettlementButton.tsx (hunk 22 above)
Comment (blazejkustra) [reply src/components/SettlementButton.tsx:218]: Please use `src/types/utils/viewRef.ts` util here and change types in both files to `View | HTMLDivElement` if possible

**Review by blazejkustra — APPROVED** [review]

**Review by hoangzinh — APPROVED** [review]

**Review by AndrewGable — DISMISSED** [review]

**Comment by AndrewGable** [comment]
Conflicts: 
Screenshot 2024-02-05 at 11 12 11 AM

**Review by hoangzinh — APPROVED** [review]
All goods. Tested as well

**Review by AndrewGable — APPROVED** [review]

**Comment by situchan** [comment]
TS check failing on main. Caused by this PR
